user_sync/rules.py

```diff
@@ -152,14 +169,45 @@ def read_desired_user_groups(self, mappings, directory_connector):
             for group in directory_user['groups']:
+                self.after_mapping_hook_scope['source_groups'].add(group) # this is a directory group name
                 dashboard_groups = mappings.get(group)
                 if (dashboard_groups != None):
                     for dashboard_group in dashboard_groups:
-                        organization_info = self.get_organization_info(dashboard_group.organization_name)
-                        organization_info.add_desired_group_for(user_key, dashboard_group.group_name)
-    
-        self.logger.info('Total directory users after filtering: %d', len(filtered_directory_user_by_user_key))        
+                        self.after_mapping_hook_scope['target_groups'].add(self.make_dashboard_group_qualified_name(dashboard_group.group_name, dashboard_group.organization_name))
+
+            # only if there actually is hook code: set up rest of hook scope, invoke hook, update user attributes
+            if (options['after_mapping_hook'] is not None):
+                self.after_mapping_hook_scope['source_attributes'] = directory_user['source_attributes'].copy()
+
+                # [TODO morr 2017-02-26]: Is there a more robust way to assemble the target attributes?
+                self.after_mapping_hook_scope['target_attributes'] = directory_user.copy()
+                self.after_mapping_hook_scope['target_attributes'].pop('groups', None)
+                self.after_mapping_hook_scope['target_attributes'].pop('identitytype', None)
+                self.after_mapping_hook_scope['target_attributes'].pop('source_attributes', None)
+
+                # invoke the customer's hook code
+                # [TODO morr 2017-02-27: Not putting any guardrails around the customer's code, which is treated as
+                # friendly. Is there anything we should be doing?
+                self.log_after_mapping_hook_scope(True)
+                exec(options['after_mapping_hook'], self.after_mapping_hook_scope)
+                self.log_after_mapping_hook_scope(False)
+                directory_user.update(self.after_mapping_hook_scope['target_attributes'])
+
+            for target_group_qualified_name in self.after_mapping_hook_scope['target_groups']:
+                target_group_name, target_organization_name = self.parse_dashboard_group_qualified_name(target_group_qualified_name)
+                target_group = Group.get_dashboard_group(target_group_name, target_organization_name)
+                if (target_group is not None):
+                    organization_info = self.get_organization_info(target_organization_name)
+                    organization_info.add_desired_group_for(user_key, target_group_name)
+                else:
+                    self.logger.error('Target dashboard group %s is not known; ignored', target_group_qualified_name)
+
+        self.logger.info('Total directory users after filtering: %d', len(filtered_directory_user_by_user_key))
```

Comment:
Is there some corresponding log of Total directory users before filtering?  If not, why not, and if so, why isn't this statement next to it?